ports/stm32/sdram.c

```diff
@@ -254,52 +253,109 @@ static void sdram_init_seq(SDRAM_HandleTypeDef
-            return false;
+    #if defined(__DCACHE_PRESENT)
+    bool i_cache_disabled = false;
+    bool d_cache_disabled = false;
+
+    // Disable caches for testing.
+    if (SCB->CCR & (uint32_t)SCB_CCR_IC_Msk) {
+        SCB_DisableICache();
+        i_cache_disabled = true;
+    }
+
+    if (SCB->CCR & (uint32_t)SCB_CCR_DC_Msk) {
+        SCB_DisableDCache();
+        d_cache_disabled = true;
+    }
+    #endif
+
+    // Test data bus
+    for (uint32_t i = 0; i < MICROPY_HW_SDRAM_MEM_BUS_WIDTH; i++) {
+        *((uint32_t *)mem_base) = (1 << i);
+        if (*((uint32_t *)mem_base) != (1 << i)) {
+            snprintf(error_buffer, sizeof(error_buffer),
+                "Data bus test failed at 0x%p expected 0x%x found 0x%lx",
+                &mem_base[0], (1 << i), ((uint32_t *)mem_base)[0]);
+            __fatal_error(error_buffer);
         }
     }
 
-    /* test address bus */
-    /* Check individual address lines */
+    // Test address bus
     for (uint32_t i = 1; i < MICROPY_HW_SDRAM_SIZE; i <<= 1) {
         mem_base[i] = pattern;
         if (mem_base[i] != pattern) {
-            printf("address bus lines test failed! address (%p)\n", &mem_base[i]);
-            return false;
+            snprintf(error_buffer, sizeof(error_buffer),
+                "Address bus test failed at 0x%p expected 0x%x found 0x%x",
+                &mem_base[i], pattern, mem_base[i]);
+            __fatal_error(error_buffer);
         }
     }
 
-    /* Check for aliasing (overlaping addresses) */
+    // Check for aliasing (overlaping addresses)
     mem_base[0] = antipattern;
     for (uint32_t i = 1; i < MICROPY_HW_SDRAM_SIZE; i <<= 1) {
         if (mem_base[i] != pattern) {
-            printf("address bus overlap %p\n", &mem_base[i]);
-            return false;
+            snprintf(error_buffer, sizeof(error_buffer),
+                "Address bus overlap at 0x%p expected 0x%x found 0x%x",
+                &mem_base[i], pattern, mem_base[i]);
+            __fatal_error(error_buffer);
         }
     }
 
-    /* test all ram cells */
-    if (!fast) {
-        for (uint32_t i = 0; i < MICROPY_HW_SDRAM_SIZE; ++i) {
+    // Test all RAM cells
+    if (exhaustive) {
+        // Write all memory first then compare, so even if the cache
+        // is enabled, it's not just writing and reading from cache.
+        // Note: This test should also detect refresh rate issues.
+        for (uint32_t i = 0; i < MICROPY_HW_SDRAM_SIZE; i++) {
             mem_base[i] = pattern;
+        }
+
+        for (uint32_t i = 0; i < MICROPY_HW_SDRAM_SIZE; i++) {
             if (mem_base[i] != pattern) {
-                printf("address bus test failed! address (%p)\n", &mem_base[i]);
-                return false;
+                snprintf(error_buffer, sizeof(error_buffer),
+                    "Address bus slow test failed at 0x%p expected 0x%x found 0x%x",
+                    &mem_base[i], pattern, mem_base[i]);
+                __fatal_error(error_buffer);
```

Comment:
We're currently using the return value from `sdram_test()` to enable/disable the sdram usage as heap but allowing the board to start up anyway, with the same pattern as used in `ports/stm32/boards/STM32F429DISC/mpconfigboard.h`
```
#define MICROPY_HEAP_START  ((sdram_valid) ? sdram_start() : &_heap_start)
#define MICROPY_HEAP_END    ((sdram_valid) ? sdram_end() : &_heap_end)
```
This then allows/requires the application code to detect that there is less than expected ram available and handle the error appropriately (in our applications startup self tests).

Changing this to `__fatal_error` would break our usage, but is probably more sensible overall. Perhaps separately I can make the `__fatal_error` function `weak` so it can be overridden in board profiles such that I can add back suitable custom handling for us.

With a `__fatal_error` change though the return value is now superfluous so should be removed, with a similar update (simplification) to remove its usage in main.c and ports/stm32/boards/STM32F429DISC/mpconfigboard.h

Reply:
I don't know how to fix this, I want to return a value too but I can't reserve a buffer specially for the error messages, maybe a change in how it's called from main can fix this, without reserving memory:
```c
bool do_sdram_test()
{
   char error_buffer[1024];
   if (sdram_test(error_buffer) == false) {
      __fatal_error(error_buffer);
   }
   return true;
}
....
void main()
{
   do_sdram_test();
   ....
}
```

Reply:
My point was that `__fatal_error` is an infinite loop / non-returning, so the `sdram_test()` return value is only ever `true` now and doesn't really add value. A sdram failure really should be considered a hardware failure so I think the deadlock `__fatal_error` is correct here (and by default will flash leds to let user know there's a problem).

Another option to pass the error messages out could be to use the `MP_ERROR_TEXT()` macro to declare the errors, they can then be passed around easily (I believe) like:
```
mp_rom_error_text_t ffs_strerror(FRESULT res)
{
    switch (res) {
        case FR_OK:
            return MP_ERROR_TEXT("Succeeded");
        case FR_DISK_ERR:
            return MP_ERROR_TEXT("A hard error occurred in the low level disk I/O layer");
        case FR_INT_ERR:
            return MP_ERROR_TEXT("Assertion failed");
        default:
            return NULL;
}
```

Reply:
That's right, I only kept the return value to avoid changing the function prototype.
> Another option to pass the error messages out could be to use the MP_ERROR_TEXT()

This is okay too, but note it will not return specific messages with addresses and expected vs actual values.

Reply:
I'm not sure how to make this return a useful error message without jumping to __fatal_error, we can probably drop the test update commits for now.

Reply:
You could even just make it such that when extended is enabled, it printf's the specific error messages? Or, yeah, just go to __fatal_error be done with it. If __fatal_error is made WEAK we can make our own overridden version of it with our own custom printf and/or led display etc.